Document the use of the HTTP01 Gateway API solver (#625)
cert-manager/website: content/en/docs/configuration/acme/http01/_index.md

type: "docs"
---

{{% pageinfo color="info" %}}

> 📌  This page focuses on solving ACME HTTP-01 challenges. If you are looking for
> on how automatically creating Certificate resources by annotating Ingress or

Reviewer: ```suggestion
📌  This page focuses on solving ACME HTTP-01 challenges. If you are looking for
how to automatically create Certificate resources by annotating Ingress or
```

overriding entries with the same key.

No other fields of the ingress can be edited.

## Configuring the HTTP01 Gateway API solver

{{% pageinfo color="info" %}}

📌  This feature requires the installation of the Gateway API CRDs. This feature
is experimental and requires a flag to be passed to the cert-manager controller:

```bash
cert-manager-controller --feature-gates=ExperimentalGatewayAPI=true

> ```

Reviewer: Should this instead be a snippet for the kube manifests?

With Helm, you can enable the experimental feature using:

```bash
helm install cert-manager jetstack/cert-manager \
  --set featureGates="ExperimentalGatewayAPI=true"
```

{{% /pageinfo %}}

Gateway API compatibility:

| Version | Compatibility |
|---------|---------------|
| v0.1.0  | ✅             |
| v0.2.0  | ✅             |
| v0.3.0  | ✅             |

If you would like to disable this feature entirely, you can remove the RBAC

> rules that allow cert-manager to access the Gateway API objects.

Reviewer: How do I do this? It also seems strange to disable a feature by changing permissions.

overriding entries with the same key.

No other fields of the ingress can be edited.

## Configuring the HTTP01 Gateway API solver

This feature, introduced in cert-manager 1.5, is currently experimental and

> subject to change.

Reviewer 2: Maybe do that same feature state thing here too.

{{% pageinfo color="info" %}}

📌  This feature requires the installation of the Gateway API CRDs. No

> additional flag is needed on the cert-manager controller; cert-manager discovers

Reviewer 2: I like this info box. Shall we also use an info box in the gateway shim docs, for consistency.

Author: Done!

overriding entries with the same key.

No other fields of the ingress can be edited.

## Configuring the HTTP-01 Gateway API solver

**FEATURE STATE**: cert-manager 1.5 [alpha]

The Gateway and HTTPRoute resources are part of the [Gateway API][gwapi], a set

> of CRDs that you install on your Kubernetes cluster and which provide various

Reviewer: ```suggestion
of CRDs that you install on your Kubernetes cluster that provide various
```

improvements over the Ingress API.

[gwapi]: https://gateway-api.sigs.k8s.io

{{% pageinfo color="info" %}}

> 📌  This feature requires to pass a feature flag to the cert-manager controller
> and the installation of the Gateway API CRDs.

Reviewer: ```suggestion
📌  This feature requires to passing a feature flag to the cert-manager controller
and the installation of the Gateway API CRDs.
```

Reviewer: We should swap installing the CRDs and feature flag in this sentence since that is the order of the walk through

To install the Gateway API CRDs, run the following command:

```sh
kubectl kustomize "github.com/kubernetes-sigs/gateway-api/config/crd?ref=v0.3.0" | kubectl apply -f -
```

To enable the feature in cert-manager, turn on the `GatewayAPI` feature gate:

- If you are using Helm:

  ```sh

>   helm upgrade --install cert-manager jetstack/cert-manager --set "extraArgs={--feature-gates=ExperimentalGatewayAPISupport=true}"

Reviewer: Namespace `cert-manager`

  ```

- If you are using the raw cert-manager manifests, add the following flag to the
  cert-manager controller Deployment:

  ```yaml
  args:
    - --feature-gates=ExperimentalGatewayAPISupport=true
  ```

The Gateway API CRDs should either be installed before cert-manager starts or
the cert-manager Deployment should be restarted after installing the Gateway API
CRDs. This is important because some of the cert-manager components only perform
the Gateway API check on startup. You can restart cert-manager with the
following command:

```sh
kubectl rollout restart deployment cert-manager -n cert-manager
```

{{% /pageinfo %}}

The Gateway API HTTPRoute HTTP-01 solver creates a temporary HTTPRoute using the
given labels. These labels must match a Gateway that contains a listener on port
80.

Here is an example of a HTTP-01 ACME Issuer using the Gateway API:

```yaml
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  name: letsencrypt

>   namespace: default

Reviewer: We are mixing namespaces (default/traefik), should we change this?

Reviewer: Ah I see now, this is talked about further down

of solving HTTP-01 challenges, but you can also choose to re-use your existing
Gateway, as long as it has a listener on port 80.

The `labels` on your Issuer may reference a Gateway that is on a separate
namespace, as long as the Gateway's port 80 listener is configured with `from:
All`. Note that the Certificate will still be created on the same namespace as
the Issuer, which means that you won't be able to reference this Secret in the
above-mentioned Gateway.

When the above Issuer is presented with a Certificate, cert-manager creates the
temporary HTTPRoute. For example, with the following Certificate:

```yaml
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: example-tls
  namespace: default
spec:
  issuerRef:
    name: letsencrypt
  dnsNames:
  - example.net
```

You will see an HTTPRoute appear. The `labels` given to this HTTPRoute are the
labels configured on the Issuer:

```yaml
kind: HTTPRoute
metadata:
  name: cm-acme-http-solver-gdhvg
  namespace: default
  labels:
    gateway: http01-solver # Copied from Issuer's `gatewayHTTPRoute.labels`.
spec:
  gateways:
    allow: All
  hostnames:

>   - traefik.mael-valais-gcp.jetstacker.net

Reviewer: We should use the same DNS as that in the Certificate

  rules:
  - forwardTo:
    - port: 8089
      serviceName: cm-acme-http-solver-gdhvg
      weight: 1
    matches:
    - path:
        type: Exact
        value: /.well-known/acme-challenge/YadC4gaAzqEPU1Yea0D2MrzvNRWiBCtUizCtpiRQZqI
```

After the Certificate is issued, the HTTPRoute is deleted.

### `labels` {#gatewayhttproute-labels}

These labels are copied into the temporary HTTPRoute created by cert-manager for
solving the HTTP-01 challenge. These labels must match one of the Gateway

> resources on your cluster. The matched Gateway have a listener on port 80.

Reviewer: What happens if no Gateways match?

Author: I added the following note:

Note that when the labels do not match any Gateway on your cluster, cert-manager will create the temporary HTTPRoute challenge and nothing will happen.

overriding entries with the same key.

No other fields of the ingress can be edited.

## Configuring the HTTP-01 Gateway API solver

**FEATURE STATE**: cert-manager 1.5 [alpha]

The Gateway and HTTPRoute resources are part of the [Gateway API][gwapi], a set
of CRDs that you install on your Kubernetes cluster that provide various
improvements over the Ingress API.

[gwapi]: https://gateway-api.sigs.k8s.io

{{% pageinfo color="info" %}}

📌  This feature requires the installation of the Gateway API CRDs and passing a
feature flag to the cert-manager controller.

To install the Gateway API CRDs, run the following command:

```sh
kubectl kustomize "github.com/kubernetes-sigs/gateway-api/config/crd?ref=v0.3.0" | kubectl apply -f -
```

To enable the feature in cert-manager, turn on the `GatewayAPI` feature gate:

- If you are using Helm:

  ```sh
  helm upgrade --install cert-manager jetstack/cert-manager --namespace cert-manager \
    --set "extraArgs={--feature-gates=ExperimentalGatewayAPISupport=true}"

>   ```

Reviewer 2: Random thought: If you'd previously enabled the experimental CertificateRequestSupport would this command disable that. I do not propose changing this command, but it is worth thinking about how easy it is to change `extraArgs` values during an upgrade.

Author: Good point, I wish we could instead re-use the same flag multiple times:

```sh
  helm upgrade --install cert-manager jetstack/cert-manager --namespace cert-manager \
    --set extraArgs="--feature-gates=ExperimentalGatewayAPISupport=true"
    --set extraArgs="--feature-gates=ExperimentalCertificateRequestSupport=true"
```

But the only way to set arrays from the command line is to use the syntax `extraArgs={...,...}` 😞
